java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/EncoderUtils.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import org.apache.arrow.util.DataSizeRoundingUtil;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.ValueVector;
+
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * Utilities for dictionnary encoding/decoding.
+ */
+public final class EncoderUtils {
+
+  /**
+   * Decodes an input vector. The algorithm takes O(n) time,
+   * where n is the length of the input vector.
+   * @param input the input vector.
+   * @param output the output vector.
+   * @param dictionary the dictionary.
+   * @param encodeNull a flag indicating if null should be decoded.
+   */
+  public static <E extends BaseIntVector, D extends ValueVector> void decode(
+          E input, D output, D dictionary, boolean encodeNull) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
```

Comment:
I still don't think I understand why encodeNull has any part in this logic.  If the input vector has a null value, we should honor that.

Reply:
Let's have some analysis over the encodeNull flag:

1. When the encodeNull is disabled:

There is no memory saving, since we are using a int vector, for which a null value also takes space.
The output element is set to null directly, and so there is no overhead for searching for the key in the dictionary.

2. When the encodeNull is enabled:

We need to search for the null element in the dictionary.
For each iteration of the encode/decode loop, there is no branch statement, and there is no need to check the validity bit. This leads to performance gain. 

What do you think? 
If you insist, we can remove it.

Reply:
my issue is encodeNull is providing should be intrinsic to the vector (whether there are any nulls are present).  So either this there should be a call to getNullCount which right now isn't O(1), so it wouldn't be great performance wise, but its the best we have.   I think it is hard to make the guarantee that the indices will all be all non-null even though you are passing in the flag.

Reply:
I see your point. Thanks for the great suggestion!

Reply:
Sorry, please hold. My current code is not correct. 

Whether we should encode a null should not depend on if there is a null in the dictionary.
An encode null flag should be enforced from outside. 

To see this, note that if there is a null in the input and there is no null in the dictionary, 
1) if encodeNull = true, we should throw an exception (the null is unexpected)
2) if encodeNull = false, we should simply set the output element to null.

I will change the code accordingly when I have a time. 
Thanks a lot.

Reply:
I don't think the encodeNull should be part of decoding.

1.  No null in dictionary, No nulls in input
2.  Nulls in dictionary, No Null in input
3. no nulls in dictionary, nulls in input.
4.  Nulls in both dictionary and input

Case 1 and 2 can be optimized (by avoiding the null check on input) but it should be empirically determined based on the null count of the Input Vector if we want to optimize, not an external flag (ideally the null count computation would be cheap).

Reply:
Suppose we do not have the external flag, when the encoder encounters a null, it has only one choice: simple output a null. 
However, this behavior can be incorrect: if the encoder has encodeNull set, it should never see a null in the input. So the correct behavior should be throwing an exception. 

The existence of null in the input vector is something we have not considered yet. This will lead to further performance improvement, because checking null existence is cheap. Can we do this in a follow up?

Reply:
I have updated the code. Sorry I did not make some points clear at first. 
In the revised code, I have added more comments to describe the encodeNull flag. And accepted some suggestions from @emkornfield . More tests cases are added. 

we have two booleans: 
encodeNull = if we should encode a null, it determines the behavior when encountering a null in the input:
if the flag is set, we should search for the null value in the dictionary, and output the index
if the flag is clear, we simply output a null

countainsNull = if there is any null in the dictionary

so we have 4 combinations:

encodeNull = true, countainsNull = true: we proceed normally, with branch removed in the encode/decode logic
encodeNull = true, countainsNull = false: we should throw an exception in the constructor.
encodeNull = false, countainsNull = true: we optimize the case by copying validity bits in batch
encodeNull = false, countainsNull = false: we proceed normally (this case can be optimized by the suggestion of @emkornfield )

Reply:
> Suppose we do not have the external flag, when the encoder encounters a null, it has only one choice: simple output a null.
However, this behavior can be incorrect: if the encoder has encodeNull set, it should never see a null in the input. So the correct behavior should be throwing an exception.

My argument is for a general purpose library we shouldn't be picky at this level.  If you have a system invariant built on top of this library, the place to do the validation is at the higher level.

Reply:
@emkornfield . I see your points and agree with your points. Thanks for the clarification. 

It seems the above discussions are getting more complicated than necessary. So I want to have a brief summary, and clear things out.

About the encodeNull flag, it simply determines encoding/decoding behaviors: should it simply outputs a null, or search for the null in the dictionary.

The benefits of turning on the encodeNull flag:
1. No need to check the validity bit in each iteration.
2. Removes the if statement in the loop

The shortcoming of turning on the encodeNull flag:
1. it searches the null value in the dictionary, which may lead to performance overhead.

So what is your suggestion? Should we keep the flag, or remove it?

Reply:
my suggestion is to not include decoding in this code at all, but centralize it at one place.  

Also, ss it stands now, it still doesn't seem like the code in the vector package is on the path to deprecation since @tianchen92 is continuing to put functionality there?

Reply:
> my suggestion is to not include decoding in this code at all, but centralize it at one place.
Sounds reasonable. I will revise the code accordingly later.

> Also, ss it stands now, it still doesn't seem like the code in the vector package is on the path to deprecation since @tianchen92 is continuing to put functionality there?

Sure. No problem. We will do it only when the timing is right.

Reply:
@emkornfield 
I have removed the encode API from the encoder class, and the EncodeUtil class is also removed.
In the tests, the decode functionality is based on DictionaryEncoder#decode static method in the vector module.